### DISCUSSION TITLE
Is there a way to implement early stopping in routing solver?

### DISCUSSION THREAD

**Discussion by jeffreyrdcs** [google/or-tools#2880] [Routing (and legacy CP) questions] [answered]
I am following the CVRP guide for python to solve a similar problem. I see that there are a few search limit options for the routing solver [https://developers.google.com/optimization/routing/routing_options](url), and in the guide it uses the time limit as an example. 

`    search_parameters.local_search_metaheuristic = (
        routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH)
    search_parameters.time_limit.FromSeconds(1)
`

Nevertheless, given the number of nodes I have, neither of those hard limits are ideal for my problem. Ideally I want to track the objective cost and stop the search early, before they hit the hard time or number of solution search limits. Is there an easy way to implement something like a custom search limit?
I saw that there is a way to do that with the CP solver [https://github.com/google/or-tools/blob/master/ortools/sat/doc/solver.md#stopping-search-early](url) but I can't see any StopSearch Method for the routing solver. 
I also found that the Solver class has a CustomLimit method, but it is unclear how it should be used with SolveWithParameters.  Is this possible with the routing solver?  Thanks!

**Comment by cryptonome** [comment, marked as answer]
Depending on what your goal is you can set a maximum processing time of e.g. 5 minutes, use a search monitor to stop after some time without improvements in the solution, etc. This has been answered multiple times, but just in case. 

```python
bestCollector = solver.BestValueSolutionCollector(False)
bestCollector.AddObjective(routing.CostVar())
routing.AddSearchMonitor(bestCollector)
bestSolution = dict(score=IntergerMaxValue, clock=0, startTime=time.time())
def NoImprovement():
  if bestCollector.SolutionCount() > 0 and bestCollector.ObjectiveValue(0) < bestSolution['score']:
    bestSolution['score'] = bestCollector.ObjectiveValue(0)
    bestSolution['clock'] = bestCollector.WallTime(0)

  return bestSolution['score'] < intergerMaxValue and \
             time.time() - bestSolution['startTime'] > \
             bestSolution['clock'] * 0.001 + 2

routing.AddSearchMonitor(solver.CustomLimit(NoImprovementLimit))
```

**Reply by jeffreyrdcs** [reply]
Thank you very much! It is working.

For future readers, whenever you try this and encounter a OverflowError in one of the IndexToNode method or some function "returned a result with an error set", it is because your CustomLimit callback (the NoImprovementLimit above) is not returning correct values in some occasion. It should only return bool.
